scripts/merge-script/index.ts

```diff
@@ -0,0 +1,120 @@
+ */
+export class PullRequestMergeTask {
+  /** Git client that can be used to execute Git commands. */
+  git = new GitClient(this._githubToken, this.config);
+
+  constructor(public config: Config, private _githubToken: string) {}
+
+  /**
+   * Merges the given pull request and pushes it upstream.
+   * @param prNumber Pull request that should be merged.
+   * @param force Whether non-critical pull request failures should be ignored.
+   */
+  async merge(prNumber: number, force = false): Promise<MergeResult> {
+    if (this._hasUncommittedChanges()) {
+      return {status: MergeStatus.DIRTY_WORKING_DIR};
+    }
+
+    const pullRequestOrFailure = await loadAndValidatePullRequest(this, prNumber, force);
+
+    if (!isPullRequest(pullRequestOrFailure)) {
+      return {status: MergeStatus.FAILED, failure: pullRequestOrFailure};
+    }
+
+    const pullRequest: PullRequest = pullRequestOrFailure;
+    const strategy = this.config.githubApiMerge ?
+        new GithubApiMergeStrategy(this.config, this.git) :
+        new AutosquashMergeStrategy(this.config, this.git);
+
+    // Branch that is currently checked out so that we can switch back to it once
+    // the pull request has been merged.
+    let previousBranch: null|string = null;
+
+    // The following block runs Git commands as child processes. These Git commands can fail.
+    // We want to capture these command errors and return an appropriate merge request status.
+    try {
+      previousBranch = this.git.getCurrentBranch();
+
+      // Run preparations for the merge (e.g. fetching branches).
+      await strategy.prepare(pullRequest);
+
+      // Perform the merge and capture potential failures.
+      const failure = await strategy.merge(pullRequest);
+      if (failure !== null) {
+        return {status: MergeStatus.FAILED, failure};
+      }
+
+      // Switch back to the previous branch. We need to do this before deleting the temporary
+      // branches because we cannot delete branches which are currently checked out.
+      this.git.run(['checkout', '-f', previousBranch]);
+
+      await strategy.cleanup(pullRequest);
+
+      // Return a successful merge status.
+      return {status: MergeStatus.SUCCESS};
+    } catch (e) {
+      // Catch all git command errors and return a merge result w/ git error status code.
+      // Other unknown errors which aren't caused by a git command are re-thrown.
+      if (e instanceof GitCommandError) {
+        return {status: MergeStatus.UNKNOWN_GIT_ERROR};
+      }
+      throw e;
+    } finally {
+      // Always try to restore the branch if possible. We don't want to leave
+      // the repository in a different state than before.
+      if (previousBranch !== null) {
+        this.git.runGraceful(['checkout', '-f', previousBranch]);
+      }
+    }
+  }
+
+  /** Gets whether the current Git repository has uncommitted changes. */
+  private _hasUncommittedChanges(): boolean {
```

Comment:
Does it make more sense for this to be a method on `GitClient`?

Reply:
Yeah, I think that makes sense. Done.